com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/verifier/JarVerifier.java

```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    private void verifyJarEntries(List<X509Certificate> targetCerts, List<JarEntry> jarEntries, String certificateCN) {
+        for (JarEntry jarEntry : jarEntries) {
+            Certificate[] entryCertificates = jarEntry.getCertificates();
+            if (entryCertificates == null || entryCertificates.length == 0) {
+                checkJarEntryName(jarEntry);
+                continue;
+            }
+            verifyJarEntryIsSigned(targetCerts, entryCertificates, certificateCN);
+        }
+    }
+
+    private void checkJarEntryName(JarEntry jarEntry) {
+        if (!jarEntry.getName()
+                     .startsWith(META_INF)) {
+            throw new SecurityException(Messages.ARCHIVE_CONTAINS_UNSIGNED_FILES);
+        }
+    }
+
+    private void verifyJarEntryIsSigned(List<X509Certificate> targetCerts, Certificate[] entryCertificates, String certificateCN) {
+        boolean signedAsExpected = isSignedAsExpected(targetCerts, entryCertificates, certificateCN);
+        if (!signedAsExpected) {
+            throw new SecurityException(Messages.ARCHIVE_IS_NOT_SIGNED_BY_TRUSTED_SIGNER);
+        }
+    }
+
+    private boolean isSignedAsExpected(List<X509Certificate> targetCerts, Certificate[] entryCertificates, String certificateCN) {
+        List<X509Certificate> entryCertificateChain;
+        int startIndex = 0;
+        while (!(entryCertificateChain = getAChain(entryCertificates, startIndex)).isEmpty()) {
+            if (validCertificateChain(targetCerts, entryCertificateChain, certificateCN)) {
+                return true;
+            }
+            startIndex += entryCertificateChain.size();
+        }
+        return false;
+    }
+
+    List<X509Certificate> getAChain(Certificate[] entryCertificates, int startIndex) {
+        if (startIndex > entryCertificates.length - 1) {
+            return Collections.emptyList();
+        }
+        int nextChainIndex = findNextChainIndex(entryCertificates, startIndex);
+        int certificateChainSize = (nextChainIndex - startIndex) + 1;
+        return getX509CertificateChain(entryCertificates, startIndex, certificateChainSize);
+    }
+
+    private int findNextChainIndex(Certificate[] certs, int startIndex) {
+        int chainIndex;
+        for (chainIndex = startIndex; chainIndex < certs.length - 1; chainIndex++) {
+            if (!Objects.equals(((X509Certificate) certs[chainIndex + 1]).getSubjectDN(),
+                                ((X509Certificate) certs[chainIndex]).getIssuerDN())) {
+                break;
+            }
+        }
+        return chainIndex;
+    }
+
+    private List<X509Certificate> getX509CertificateChain(Certificate[] certificates, int startIndex, int certChainSize) {
+        List<X509Certificate> x509Certificates = new ArrayList<>();
+        for (int i = 0; i < certChainSize; i++) {
+            x509Certificates.add((X509Certificate) certificates[startIndex + i]);
+        }
+        return x509Certificates;
```

Comment:
This could be substituted for IntStream.range(0, certChainSize).mapToObj().collect()

Reply:
Done. Replaced